### PR TITLE
Curtain plot (model) with Observations overlaid as scatter & Updated Violin Plot with bigger label fonts and ticks consistent with other plots

### PR DESCRIPTION
**Pull request by quaz115**
image

### CODE REVIEW

**Review by rschwant — COMMENTED** [review]
Can you fix the plot title to be based on text_kwargs too. To make it look nice I also had the tickparams a bit smaller than the title and x/y labels at 0.8*. You can look at make_boxplot subroutine in surfplots.py for how to do this.

**Comment by quaz115** [comment]
> Can you fix the plot title to be based on text_kwargs too. To make it look nice I also had the tickparams a bit smaller than the title and x/y labels at 0.8*. You can look at make_boxplot subroutine in surfplots.py for how to do this.

image
Made this fix in the make_violin_plot() in aircraftplots.py  (see https://github.com/NOAA-CSL/MELODIES-MONET/pull/254/commits/5f32335c3b70303bcd2d3a3ab93122daf01ea8a2)

**Comment by quaz115** [comment]
Curtain plot added:

The **interpolate_and_clean_model_data function** is responsible for preparing the model data for visualization, which involves creating a meshgrid from the time and altitude data, interpolating over any gaps, and cleaning the dataset by replacing any non-numerical values with appropriate limits. The **make_curtain_plot function** takes this prepared data and produces the actual visualization, plotting the model data as a contour plot (a curtain plot) and overlaying observational data as a scatter plot on top of this. The function uses the matplotlib library for plotting, allowing for a custom (color_map_custom as True) or predefined colormap (color_map_custom as False), and ensures the plot is correctly normalized using the minimum and maximum data values. The appearance of the plot(s) (e.g., labels, titles) are customized using the parameters specified in the YAML configuration.

Currently, it prints for each model-obs pair as done for scatter plots earlier. For example:
 ```
Saving Curtain plot (Model) with Scatter Overlay (Observation) to /wrk/qrasool/output_testaircraft_PR189_uptoviolin_multiplemodels/plot_grp5.curtain.O3_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS_firexaq_vs_wrfchem_v4.2.png...
Processing curtain plot for model 'wrfchem_v4.2_test' and observation 'firexaq'...
```

image

**Comment by quaz115** [comment]
Also @rschwant the test data used here is for FIREX 2019-09-05, I think model results might be a bit old for this: /wrk/qrasool/firex_mech_qzhu/run_CONUS_fv19_BEIS_1.0xISO_RACM_v4.2.2_racm_berk_vcp_noI_phot_soa/0905/* (on Meiyu)

**Comment by quaz115** [comment]
@colin-harkins @rschwant For the latest [commits](https://github.com/NOAA-CSL/MELODIES-MONET/pull/254/commits/0285d6a1f7057e042e90decde21249be43e456e6) to fix curtain plot, i am able to generate model curtain plot (pressure as y-axis and time as x-axis) with model scatter as overlay aligned with obs scatter (in the second subplot). Curtain and scatter profiles all share common colorbar (vmin and vmax can be user-defined or dynamically set based on the data min max) for each variable set:   An example here:  a) using default Spectral_r cmap:
image    
image    
image

      b) using user-defined custom_color_map in YAML set to True:  
image
 
image

image

File: melodies_monet/plots/aircraftplots.py
Comment (rschwant) [inline melodies_monet/plots/aircraftplots.py:368]: Can you save the plot too? Like you did for the violin plot?

**Review by rschwant — COMMENTED** [review]
Great job, pulling in a very complicated plot! I had a couple questions still. I'll be in the office tomorrow if you want to talk about my comments. If you can just address these this week that works well we can work on the spatial plot another time.

**Comment by rschwant** [comment]
Also I just noticed can you make sure the vmin and vmax for the pressure on the y-axis for both plots are the same so it is easier to compare them?

File: melodies_monet/driver.py
Comment (colin-harkins) [reply melodies_monet/driver.py:1703]: @rschwant In the pairing step the full vertical profile at matched horizontal coordinates is not saved in the paired data so it needs to be made again if we keep the same paired data format. The options here were to somehow create a whole new different pairing object in the pair class instance just for curtain plots or to remake the paired data with the full vertical profile. In my opinion this is easier and more simple to just remake the vertical info when plotting curtain plots than fundamentally change the structure of the pair class.

Comment (rschwant) [reply melodies_monet/driver.py:1703]: I definitely agree with remaking the vertical profile since this is not saved and that is used to create the contour plot in the vertical. I agree we do not want to save this in the paired netcdf files. But I'm still confused, for the markers overlaid on top of both subplot 1 and 2, why can you not just use the paired data for the model in subplot 1 and the observational data in subplot 2? Maybe we can meet on Monday to discuss. Maybe I am misunderstanding something.

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Removed this as not used currently

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: Resolved

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: vmin , vmax as used for altitude variable limits as y-axis used (same as for vertprofile plt_grp)

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: Edited as suggested

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: dynamic_color_scale removed

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: Deleted

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: Corrected

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: Removed  "use_interpolation" option, since not used

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: Removed LogNorm function, since not used

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py:389]: Resolved in this `[commit](https://github.com/NOAA-CSL/MELODIES-MONET/pull/254/commits/21b893789b1d8cd89a96725bfc6f5de556bbd499)`. @colin-harkins thanks for resolving this!

Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py:389]: image

Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py:389]: image

**Comment by quaz115** [comment]
> Also I just noticed can you make sure the vmin and vmax for the pressure on the y-axis for both plots are the same so it is easier to compare them?

Lines 373-375 in aircraftplots.py now addresses this issue:    # Set the same y-tick labels for both subplots and exclude the minimum value
    y_ticks = axs[0].get_yticks()
    y_lim = axs[0].get_ylim()
    axs[1].set_yticks(y_ticks)
    axs[1].set_ylim(y_lim)

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: set more explicitly now:     ` # Handling custom color maps
    if 'color_map_custom' in grp_dict and grp_dict['color_map_custom'] is True and 'colors' in grp_dict:
        # Custom color map logic
        colors = grp_dict['colors']
        cmap = LinearSegmentedColormap.from_list("custom_cmap", colors, N=grp_dict.get('color_levels', 100))
    else:
    # Default color map logic
        cmap = plt.get_cmap(grp_dict.get('color_map', 'viridis'))`

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Removed this part as it was redundant and not needed

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: removed forced color bar minima to be 0

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (quaz115) [reply examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml]: Now,  vmin_plot and vmax_plot defined in the observations section of the yaml file are used here instead. Removed the whole 'varible_limits' option earlier in YAML and made associated changes in the driver.py and aircrfraftplots.py curtain parts

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py:1751]: vmin vmax are user defined as othe rplots (say vertprofile) for y-axis (altitude variable here). vmin_plot and vmx_plot from obs section of yaml are used for colorbar min max for plotted variabels (say, NO, NO2, O3 etc.)

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: Done

**Comment by quaz115** [comment]
Latest output example (note how changed y-axis target pressure levels (y-axis) to 500 via YAML now):   Pressure MIN:<xarray.DataArray 'pressure_model' ()> Size: 8B
array(5582.80372111), max: <xarray.DataArray 'pressure_model' ()> Size: 8B
array(101160.20015191), interval: 500
Saving curtain plot to /wrk/qrasool/output_testaircraft_PR189_uptoviolin_multiplemodels_racmesrlvcpFTPnew/plot_grp5.curtain.O3_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS_firexaq_vs_wrfchem_v4.2.png...
time shape: (1901,), ndims: 1
altitude shape: (500,), ndims: 1
model_data_2d shape: (1901, 500), ndims: 2
pairdf shape: (1901, 648), ndims: 2
cmin and cmax set dynamically based on model and observation data to 0 and 100
Saving curtain plot to /wrk/qrasool/output_testaircraft_PR189_uptoviolin_multiplemodels_racmesrlvcpFTPnew/plot_grp5.curtain.O3_CL_RYERSON.2019-09-05_12.2019-09-06_00.all.CONUS_firexaq_vs_wrfchem_v4.2.png...      
image

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Removed

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Using pairdf now for scatter overlays

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Using "pairdf" for model and obs scatter overlay now

File: melodies_monet/plots/aircraftplots.py
Comment (quaz115) [reply melodies_monet/plots/aircraftplots.py]: Done, see updated YAML and make_curtain_plot() in aircraftplots

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Now using pairdf as suggested

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: So far these are essential for model curtain 2d data generation, not sure how i can use as they are in pair_data function where pairing is done for the first time for 'aircraft' case?

Comment (quaz115) [reply melodies_monet/driver.py]: But the additional Aligning part is removed now

File: melodies_monet/driver.py
Comment (quaz115) [reply melodies_monet/driver.py]: Resolved, obs_data and model_data not used anymore

File: melodies_monet/driver.py
Comment (rschwant) [inline melodies_monet/driver.py:1208]: Do you use all of these? I looks like numpy is already pulled in at the top. mdates is not used and vert_interp is not used.

File: melodies_monet/driver.py
Comment (rschwant) [inline melodies_monet/driver.py:1702]: obs_pressure and pairdf are in a different order here than in the make_curtain_plot routine in the aircraftplots.py file and several of the ones below are also in a different order. Can you make this consistent between these two files? It makes it easier keep things consistent in the future.

File: examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml
Comment (rschwant) [inline examples/yaml/control_wrfchem_aircraft_Latestfor_develop_aircraft_testmultiplemodelgroups.yaml:270]: It does not look like these vmin and vmax values are used in the code right now. Do you want to correct this or just take this option out?

File: melodies_monet/plots/aircraftplots.py
Comment (rschwant) [inline melodies_monet/plots/aircraftplots.py:295]: The vmin and vmax values from your yaml file are not linked to the vmin and vmax values used here and this code is not used later on either. So likely should delete this line. And fix this connection like specified above.

**Review by rschwant — COMMENTED** [review]
Just had a couple small comments to fix for the vmin, vmax, cmin, and cmax variables and everything else looks great! Thanks for updating everything. Once these things are fixed, I can fully approve and we can merge this in.

**Review by colin-harkins — APPROVED** [review]
I think once you make the changes Becky requested this looks good. I'll approve on my end.

**Comment by colin-harkins** [comment]
I thought this would take two approvals to be able to merge so hold off on merging until you address Becky's comments but I'll leave as is with the approval on my end.

**Comment by quaz115** [comment]
@rschwant @colin-harkins Tested with fixes for previous comments on y-axes and colorbar labels. Tested with changing units of pressure (pressure_units via yaml, distinguished between 'intervals' as y-axis tick intervals and 'num_levels' as interpolation levels for target pressures as well). Updated single model yaml as well with curtain plot config options.  An example result with customized y-axis limits working well:   
image

However getting this CI/Check docs build error, which I don't think is related to my changes.

**Comment by colin-harkins** [comment]
It looks like there are broken links in the supported datasets page and the developers guide page on the readthedocs.

```
(develop/developers_guide: line  175) broken    https://github.com/NOAA-CSL/MELODIES-MONET/projects/2 - 404 Client Error: Not Found for url: https://github.com/NOAA-CSL/MELODIES-MONET/projects/2
(users_guide/supported_plots: line    8) broken    https://github.com/NOAA-CSL/MELODIES-MONET/projects/3 - 404 Client Error: Not Found for url: https://github.com/NOAA-CSL/MELODIES-MONET/projects/3
(users_guide/supported_datasets: line   58) redirect  https://github.com/NOAA-CSL/MELODIES-MONET/projects/4 - permanently to https://github.com/orgs/NOAA-CSL/projects/7
(users_guide/supported_datasets: line   33) ok        https://aeronet.gsfc.nasa.gov/
(users_guide/supported_datasets: line   34) redirect  http://vista.cira.colostate.edu/Improve/ - permanently to https://vista.cira.colostate.edu/Improve/
(users_guide/supported_datasets: line   64) broken    https://github.com/NOAA-CSL/MELODIES-MONET/projects/6 - 404 Client Error: Not Found for url: https://github.com/NOAA-CSL/MELODIES-MONET/projects/6
(users_guide/supported_datasets: line   44) broken    https://github.com/NOAA-CSL/MELODIES-MONET/projects/7 - 404 Client Error: Not Found for url: https://github.com/NOAA-CSL/MELODIES-MONET/projects/7
(users_guide/supported_datasets: line   23) broken    https://github.com/NOAA-CSL/MELODIES-MONET/projects/8 - 404 Client Error: Not Found for url: https://github.com/NOAA-CSL/MELODIES-MONET/projects/8
```

I think potentially this is related to some restructuring of the project board link? The project board that exists now seems to be: https://github.com/orgs/NOAA-CSL/projects/6

I'm not sure if this is related to something we changed or github changed about the project board. @rschwant do you know?

Regardless, hopefully it shouldn't be too hard to find new links for these items in the project board.

**Comment by rschwant** [comment]
Yes, I asked Quazi to update these links. Github got rid of their classic project boards, but luckily we had already migrated to one project board, so we just need to update the ReadTheDocs for this.

**Comment by quaz115** [comment]
@colin-harkins i did try to resolve conflicts but not sure why they are still visible on the PR?

**Review by rschwant — APPROVED** [review]
Looks great!

**Pull request merged**
